sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/expressions/Expressions.java

```diff
@@ -0,0 +1,141 @@
+  public Transform apply(String name, Expression... args) {
+    return LogicalExpressions.apply(name, args);
+  }
+
+  /**
+   * Create a named reference expression for a column.
+   *
+   * @param name a column name
+   * @return a named reference for the column
+   */
+  public NamedReference column(String name) {
+    return LogicalExpressions.reference(name);
+  }
+
+  /**
+   * Create a literal from a value.
+   *
+   * @param value a value
+   * @param <T> the JVM type of the value
+   * @return a literal expression for the value
+   */
+  public <T> Literal<T> literal(T value) {
+    return LogicalExpressions.literal(value);
+  }
+
+  /**
+   * Create a bucket transform for one or more columns.
+   * <p>
+   * This transform represents a logical mapping from a value to a bucket id in [0, numBuckets)
+   * based on a hash of the value.
+   *
+   * @param numBuckets the number of output buckets
+   * @param columns input columns for the bucket transform
+   * @return a logical bucket transform
+   */
+  public Transform bucket(int numBuckets, String... columns) {
+    return LogicalExpressions.bucket(numBuckets, columns);
+  }
+
+  /**
+   * Create an identity transform for a column.
+   * <p>
+   * This transform represents a logical mapping from a value to itself.
+   *
+   * @param column an input column
+   * @return a logical identity transform
+   */
+  public Transform identity(String column) {
+    return LogicalExpressions.identity(column);
+  }
+
+  /**
+   * Create a year transform for a timestamp or date column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp or date to a year, such as 2018.
+   *
+   * @param column an input timestamp or date column
+   * @return a logical year transform
+   */
+  public Transform year(String column) {
+    return LogicalExpressions.year(column);
+  }
+
+  /**
+   * Create a month transform for a timestamp or date column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp or date to a month, such as
+   * 2018-05.
+   *
+   * @param column an input timestamp or date column
+   * @return a logical month transform
+   */
+  public Transform month(String column) {
+    return LogicalExpressions.month(column);
+  }
+
+  /**
+   * Create a date transform for a timestamp or date column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp or date to a date, such as
+   * 2018-05-13.
+   *
+   * @param column an input timestamp or date column
+   * @return a logical date transform
+   */
+  public Transform date(String column) {
+    return LogicalExpressions.date(column);
+  }
+
+  /**
+   * Create a date and hour transform for a timestamp column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp to an hour, such as 2018-05-13,
+   * hour 11.
+   *
+   * @param column an input timestamp column
+   * @return a logical date and hour transform
+   */
+  public Transform dateHour(String column) {
```

Comment:
this combo seems pretty weird. i think a better way to do it is to support multiple partition levels, or use a struct to allow arbitrary composition.

Reply:
This is an hour ordinal that doesn't reset at the start of every day. With partition transforms, it makes no sense to partition by both date and hour-of-day. That just makes the implementation more complex.

I think the confusion here is in the name. What if these were named `years`, `months`, `days`, and `hours` instead?

Reply:
I don't know if I was confused earlier or am confused now. Let's say for timestamp 2019-04-02 10:11:12, what should dateHour return?

Reply:
`dateHour` doesn't return anything, it is not a concrete transformation. Instead, it is the granularity of a time partition. There are many concrete transformations that implement hourly partitioning.

Whether the storage implementation uses `{date=2019-04-02, hour=10}` or `hour=2019-04-02-10` doesn't matter. Iceberg defines this function as "Extract a timestamp hour, as hours from 1970-01-01 00:00:00", and uses an hour ordinal stored as an int.

That's why I think this may be more clear if we updated the transform names. `PARTITION BY (category, hours(ts))` is more clear that this is not a concrete transform.

Eventually, we will need to communicate what the concrete implementation of this function is. A table may report it is partitioned by `days(ts)` and Spark will need to produce those values for dataframe partitioning and write optimization. In that case, Spark should use the table's catalog to look up the function.

Reply:
We discussed this in the sync last night. There were to conclusions. First, the approach makes sense, and second, the names should be improved. We came up with `perHour`, `hourly`, and `hours`. I think `hours` makes the most sense when reading SQL so I'll update these to use plural names.